### DISCUSSION TITLE
MouseScroll as an Excel Add-in

### DISCUSSION THREAD

**Discussion by don922** [cristianbuse/VBA-UserForm-MouseScroll#38] [General]
For anyone interested you can make MouseScroll into an Excel Add-In.  I have done it and so far it is working great. The benefits of doing this is that when Christian updates the code you only need to update the add-in and all workbooks using MouseScroll will automatically be using the new code. 

You do have to make one change to MouseScroll. You need to remove or comment out the line that says 'Option Private Module'. If you don't do this then the EnableMouseScroll and DisableMouseScroll routines will not be found by your project. It took me a while to figure this out. 

If anyone is interested in doing this and needs detailed steps just let me know and I'll post them. 

Don

**Comment by cristianbuse** [comment]
Hi @don922 ,

Apologies you wasted time on this. I obviously tested the Add-In approach but forgot to mention to remove the ```Option Private Module``` line. I have now removed it myself from the repo in the last commit 6f3d9df73b292626b6c1ffb6c61fce3d163926e7

Many thanks!

**Comment by don922** [comment]
No worries. I like tracking down bugs, although this wasn’t a bug. I didn’t know that option existed so I learned something new. 

Thanks
Don

**Comment by cristianbuse** [comment]
#43 also falls under this discussion. Should the MouseScroll library handle concurrent documents if each project implements the library separately?

**Reply by don922** [reply]
If each project implements the library separately, not using it as an add-in, then no. In my add-in if I make mousescroll a class the then answer is no also. As I have it now mousescroll is a bas file and not a class and I realized that while it works, it won't work if multiple projects are using it at the same time because you have some static variables defined in some procedures. If I make mousescroll a class and then call mousescrollclass.enablescroll it will work. This is why I started another discussion on the public constants you have defined. Maybe I should have just asked here but I already started it.

**Reply by cristianbuse** [reply]
Hi Don,

> If each project implements the library separately, not using it as an add-in, then no.

I agree.

I would fail to see the benefit when working in Excel - using an add-in would be better and would make more sense and show clear intent.

For apps that do not allow an add-in, it's still probably better to implement the code for each document separately. Otherwise, documents would become dependant on each other.

**Comment by Kira-Gao** [comment]
Hello, during the actual use of the UserForm, I encountered a frustrating issue. When the user's display is in multi-screen and extended mode, the UserForm does not appear in the center of the screen as expected (with Excel maximized). The calculation of Application.Width and .Top is quite confusing.
I have tried many times but couldn't achieve the desired result. Could you please take a look and try to solve this issue?
Thank you very much for your help.

**Reply by JamieG-UK** [reply]
We've had to solve this with a dozen complex WinAPIs in our free [BrightSlide](https://brightcarbon.com/BrightSlide) add-in for PowerPoint. But this isn't related to this repo .

**Comment by cristianbuse** [comment]
Hi @Kira-Gao 

Are you referring to the demo form in this repository? Or any form in general?

**Reply by Kira-Gao** [reply]
Any form in general

**Reply by cristianbuse** [reply]
I cannot replicate the issue. I've tried using different monitor configurations and scaling and the following still works fine for me:
```VBA
Private Sub UserForm_Initialize()
    Me.StartUpPosition = 0
    Me.Left = Application.Left + Application.Width / 2 - Me.Width / 2
    Me.Top = Application.Top + Application.Height / 2 - Me.Height / 2
End Sub
```
You might want to try adjusting the ```StartUpPosition``` ar design time, not at runtime. See [this article](https://bettersolutions.com/vba/userforms/positioning.htm) for more details.
